### PR TITLE
add half cauchy distribution

### PR DESCRIPTION
**Pull request by jpchen**
not finalized. i used some of the previous distribution PRs as a reference

Implementation:
i reused the cauchy sampler and enforced positivity. the abs() is added to the computation graph, since the whole sampler is differentiable, but im not sure if this is correct. if the transformation happens in place (thereby omitting it from the graph), then the scale gradients will match the signs of the underlying cauchy sample, otherwise the location gradients will.

Tests:
tests are reused from Cauchy sans gradient tests

### CODE REVIEW

**Comment by jpchen** [comment]
cc @fritzo  (i cant add reviewers)

**Comment by alicanb** [comment]
Right now it seems that support is positive. Do we need this to fold across `loc`? or simply origin?

File: test/test_distributions.py [hunk 1]
```diff
@@ -1316,8 +1338,11 @@ def test_one_hot_categorical_shape(self):
         self.assertEqual(dist.log_prob(dist.enumerate_support()).size(), torch.Size((2, 3)))
         self.assertEqual(dist.log_prob(torch.ones((3, 1, 2))).size(), torch.Size((3, 3)))
 
-    def test_cauchy_shape_scalar_params(self):
-        cauchy = Cauchy(0, 1)
+    def test_cauchy_shape_scalar_params(self, is_half_cauchy=None):
+        if is_half_cauchy:
```
Comment (fritzo) [inline test/test_distributions.py]: It's cleaner to copy the method than to introduce this complex testing logic.

File: test/test_distributions.py [hunk 2]
```diff
@@ -514,6 +520,22 @@ def test_cauchy(self):
...
+        loc_1d = Variable(torch.zeros(1), requires_grad=True)
+        scale_1d = Variable(torch.ones(1), requires_grad=True)
+        self.assertEqual(HalfCauchy(loc, scale).sample().size(), (5, 5))
+        self.assertEqual(HalfCauchy(loc, scale).sample_n(7).size(), (7, 5, 5))
+        self.assertEqual(HalfCauchy(loc_1d, scale_1d).sample().size(), (1,))
+        self.assertEqual(HalfCauchy(loc_1d, scale_1d).sample_n(1).size(), (1, 1))
+        self.assertEqual(HalfCauchy(0.0, 1.0).sample_n(1).size(), (1,))
+
+        set_rng_seed(1)
+        self._gradcheck_log_prob(Uniform, (loc, scale))
```
Comment (fritzo) [inline test/test_distributions.py]: Replace `Uniform` with `HalfCauchy`. If you feel like cleaning up, also fix this error in `test_cauchy()`.

File: torch/distributions/half_cauchy.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+         0.3141
+        [torch.FloatTensor of size 1]
+
+    Args:
+        loc (float or Tensor or Variable): lower bound of the distribution.
+        scale (float or Tensor or Variable): half width at half maximum.
+    """
+
+    params = {'loc': constraints.real, 'scale': constraints.positive}
+    support = constraints.positive
```
Comment (fritzo) [inline torch/distributions/half_cauchy.py]: Replace this with a property below `__init__()`. See `Uniform` for example.
```py
@constraints.dependent_property
def support(self):
    return constraints.greater_than(self.loc)
```

File: torch/distributions/half_cauchy.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        else:
+            batch_shape = self.loc.size()
+        super(HalfCauchy, self).__init__(batch_shape)
+
+    def rsample(self, sample_shape=torch.Size()):
+        sample = self._cauchy.rsample(sample_shape)
+        return sample.abs()
+
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
```
Comment (fritzo) [inline torch/distributions/half_cauchy.py]: Why not correctly handle points outside of the support?
```
result = self._cauchy.log_prob(value) + math.log(2)
result[value < self.loc] = -float('inf')
return result
```
and no need to `_validate_log_prob_arg`, since that would be done inside `self._cauchylog_prob()`

File: torch/distributions/half_cauchy.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    r"""
+    Half-Cauchy distribution.
+
+    This is a continuous distribution with lower-bounded domain (`x > mu`).
+    See also the `Cauchy` distribution.
+
+    Example::
+
+        >>> m = HalfCauchy(torch.Tensor([0.0]), torch.Tensor([1.0]))
+        >>> m.sample()  # sample from a Cauchy distribution with loc=0 and scale=1
```
Comment (fritzo) [inline torch/distributions/half_cauchy.py]: Remove this line.

File: test/test_distributions.py [hunk 6]
```diff
@@ -1326,8 +1351,11 @@ def test_cauchy_shape_scalar_params(self):
         self.assertEqual(cauchy.log_prob(self.tensor_sample_1).size(), torch.Size((3, 2)))
         self.assertEqual(cauchy.log_prob(self.tensor_sample_2).size(), torch.Size((3, 2, 3)))
 
-    def test_cauchy_shape_tensor_params(self):
-        cauchy = Cauchy(torch.Tensor([0, 0]), torch.Tensor([1, 1]))
+    def test_cauchy_shape_tensor_params(self, is_half_cauchy=None):
+        if is_half_cauchy:
```
Comment (fritzo) [inline test/test_distributions.py]: ditto: favor consistency over logic

File: test/test_distributions.py [hunk 7]
```diff
@@ -119,6 +119,12 @@
             'scale': Variable(torch.randn(1).abs(), requires_grad=True),
         },
     ]),
+    Example(HalfCauchy, [
```
Comment (fritzo) [inline test/test_distributions.py]: Could you add some examples with nonzero loc, e.g. `loc=-1`?

**Comment by jpchen** [comment]
@alicanb thats the desired behavior, half cauchy is only defined for values > loc

File: test/test_distributions.py (hunk 7 above)
Comment (jpchen) [reply test/test_distributions.py]: will do

File: torch/distributions/half_cauchy.py (hunk 5 above)
Comment (jpchen) [reply torch/distributions/half_cauchy.py]: done

File: torch/distributions/half_cauchy.py (hunk 3 above)
Comment (jpchen) [reply torch/distributions/half_cauchy.py]: what does the decorator mean? how is this different than say `support = constraints.greater_than(self.loc)`

File: test/test_distributions.py (hunk 1 above)
Comment (jpchen) [reply test/test_distributions.py]: code duplication :(

File: test/test_distributions.py (hunk 2 above)
Comment (jpchen) [reply test/test_distributions.py]: done

File: torch/distributions/half_cauchy.py (hunk 4 above)
Comment (jpchen) [reply torch/distributions/half_cauchy.py]: true

File: torch/distributions/half_cauchy.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/half_cauchy.py]: Class variables cannot depend on self, hence the need to define a property. The fancy `@constraints.dependent_property` decorator is just like `@property` but adds a wrapper class so that the unbound property (e.g. `HalfCauchy.support`) is treated as a `constraints.dependent` but when bound to an instance (e.g. `HalfCauchy(0,1).support`) it is treated as a `constraints.greater_than(...)`. This pattern allows us to use cheaper static checks on most classes and use more expensive dynamic checks on classes with truly dependent properties.

If `HalfCauchy` is never used in practice with nonzero `loc`, you might consider dropping the `loc` parameter so that `.support` can stay static.

File: test/test_distributions.py (hunk 1 above)
Comment (fritzo) [reply test/test_distributions.py]: cost of abstraction :(

**Comment by fehiepsi** [comment]
@jpchen Any news on this? I guess it is almost finished. It would be great to use this distribution. :)
